packages/foundry/contracts/prediction-market/lib/PredictionMarketLib.sol

```diff
@@ -0,0 +1,297 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { 
+    PredictionMarket,
+    Side 
+} from '../Types.sol';
+
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+library PredictionMarketLib { 
+
+    /**
+     * @notice Prediction market has zero token balance for one of both pair tokens
+     */
+    error ZeroTokenBalance();
+
+    /**
+     * @notice Prediction market has insufficent liquidity to execute swap
+     */
+    error CannotSwapInMarket();
+
+    /**
+     * @notice Prediction market does not have liquidity on both sides
+     * @dev Occurs when a user tries to add liquidity proportionally to a single sided market
+     * In this case, the market cannot quote the side with zero liquidity
+     */
+     error CannotAddProportionalLiquidityToSingleSidedMarket();
+
+    
+    /**
+     * @notice Quote the current prices (probabilities) of the prediction market
+     * @dev The total value of the market is always equal to the deposited liquidity, thus to
+     * get the probability or price of each outcome, we need to divide the net outcome balance
+     * by the total deposited liquidity.
+     * @param self The prediction market to quote
+     */
+    function quote(
+        PredictionMarket memory self
+    ) internal pure returns (uint256 quoteBull, uint256 quoteBear) {
+        uint256 liquidity = netLiquidity(self);
+
+        quoteBull = Math.mulDiv(self.balanceBull, 1e18, liquidity);
+        quoteBear = Math.mulDiv(self.balanceBear, 1e18, liquidity);
+    }
+
+    /**
+     * @notice Quote the asset pair in the underlying pool
+     * @dev Uses the total assets held by the pool and provide a simple quote based on assetA relative 
+     * to assetB instead of using the pool pricing model
+     *
+     * Reverts if the pool balance of either asset is 0
+     * @param self The prediction market to quote
+     * @param vault The vault to get token balances from
+     */
+    function quoteUnderlying(
+        PredictionMarket memory self,
+        IVault vault
+    ) internal view returns(uint256) {
+        // Fetch token balances from pool for each token in the pair. 
+        (IERC20[] memory tokens,,,uint256[] memory lastBalancesLiveScaled18) = vault.getPoolTokenInfo(self.pool);
+
+        uint256 balanceToken0;
+        uint256 balanceToken1;
+
+        for(uint256 i=0; i < tokens.length; i++){
+            address token = address(tokens[i]);
+
+            if(token == self.token0) {
+                balanceToken0 = lastBalancesLiveScaled18[i];
+            } else if(token == self.token1) {
+                balanceToken1 = lastBalancesLiveScaled18[i];
+            }
+        }
+
+        // Revert if either token has a zero balance.
+        if(balanceToken0 * balanceToken1 == 0){
```

Comment:
These can't be zero if the pool is initialized. Could check when creating the market that the pool has been initialized.